framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -188,7 +188,27 @@ private static String formatAnnotations(Collection<? extends Annotation> annos)
      */
     private static String formatArrayType(ATypeElement e, String arrayType) {
         StringBuilder result = new StringBuilder();
-        return formatArrayTypeImpl(e, arrayType, result);
+        int componentEndPos;
+        if (arrayType.startsWith("@")) {
+            // if there's an explicit annotation, go to second space
```

Comment:
What if there are two annotations, as in `@Nullable @Initialized`?  What if there is a space in the annotation name, as in `@KeyFor({"a", "b"})`?

Reply:
Annotations from other hierarchies were a problem. I've added tests for both (including tests for non-array types, which also handled annotations from irrelevant hierarchies incorrectly), and fixed the issue. That required me to modify how `-AmergeStubsWithSource` works - previously, it was replacing the whole type from source with the type in the stub file if the stub file had any non-default type, which would override the original types in other hierarchies. I've corrected it to do what the documentation actually says it will (which is take the GLB for each hierarchy).

In the process, I renamed `AnnotatedTypeMerger` to `AnnotatedTypeReplacer` to be consistent with its actual function - that was what `-AmergeStubsWithSource` was using previously that produced the wrong behavior. Right now this PR includes that change, but I'm going to pull that out before I merge this into #2871 and make a separate PR against the CF.